In [2]:
import numpy as np
import matplotlib.pyplot as plt

### Упражнение 1


На одном графике изобразите время затраченное на вычисление с использованием массивов numpy и стандартных списков в питоне.
Для более точных результатов замеряйте время выполнения функции несколько раз и усредняйте.
* Используйте функцию z = 2*x**2 + 4*y
* Перемножение матриц размера n на n

Для начала инициализирую функции и переменные с необходимыми для расчётов массивами.

In [35]:
def problem_1_function_lists(X,Y):
    return [(2*x**2 + 4*y) for x,y in zip(X,Y)]
def problem_1_function_ndarrays(x,y):
    return 2*x**2 + 4*y

In [42]:
n = 1000
python_list_matrix_x = [[np.random.rand() for i in range(n)] for i in range(n)]
python_list_matrix_y = [[np.random.rand() for i in range(n)] for i in range(n)]
numpy_array_matrix_x = np.copy(python_list_matrix_x)
numpy_array_matrix_y = np.copy(python_list_matrix_y)

python_list_x = [np.random.rand() for i in range(n)]
python_list_y = [np.random.rand() for i in range(n)]
numpy_array_x = np.copy(python_list_x)
numpy_array_y = np.copy(python_list_y)

Проверю, что типы данных корректны.

In [37]:
type(python_list_matrix)

list

In [38]:
type(numpy_array_matrix)

numpy.ndarray

Использую магическую функцию timeit для расчёта времени операций.

In [39]:
%timeit problem_1_function_lists(python_list_x, python_list_y)

233 µs ± 9.49 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [40]:
%timeit problem_1_function_ndarrays(numpy_array_x, numpy_array_y)

8.19 µs ± 634 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


В случае с расчётом функции $2x^2 + 4y$ , где x и y - это элементы соответствующих массивов, которые итеративно подставляются в функцию для расчёта элемента z, мы получили большую разницу во времени выполнения расчётов для разных типов массивов.

Так, массив numpy выполнил данную функцию за 8,19 микросекунд, а массив в виде стандартного списка python выполнил ту же функцию за гораздо большее время - 233 микросекунды, что с учётом погрешности в 25-32 раза дольше.

Далее мы замерим выполнение операции умножения квадратных матриц n-ого порядка. Для списков я использую функцию ниже. Для массивов numpy возьму оператор @

In [41]:
def square_matrix_multiplication(X,Y):
    length = len(X) 
    result_matrix = [[0 for i in range(length)] for i in range(length)]
    for i in range(length):
        for j in range(length):
            for k in range(length):
                result_matrix[i][j] += X[i][k] * Y[k][j]
    return result_matrix

In [43]:
%timeit square_matrix_multiplication(python_list_matrix_x, python_list_matrix_y)

4min 52s ± 15.8 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [44]:
%timeit numpy_array_matrix_x @ numpy_array_matrix_y

33.4 ms ± 319 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


Разница коллосальная - для матриц размерностью 1000 * 1000 в случае со списками операция заняла практически 5 минут, когда в numpy массиве 33 миллисекунды. 

### Упражнение 2

* Создайте массив чисел от 2 до 75. Выведите только нечётные.
* Присвойте нечётным числам этого массива значение -1.


In [3]:
ex2_array = np.arange(2,76)
print(ex2_array[ex2_array % 2 != 0])
ex2_array[ex2_array % 2 != 0] = -1
ex2_array

[ 3  5  7  9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49
 51 53 55 57 59 61 63 65 67 69 71 73 75]


array([ 2, -1,  4, -1,  6, -1,  8, -1, 10, -1, 12, -1, 14, -1, 16, -1, 18,
       -1, 20, -1, 22, -1, 24, -1, 26, -1, 28, -1, 30, -1, 32, -1, 34, -1,
       36, -1, 38, -1, 40, -1, 42, -1, 44, -1, 46, -1, 48, -1, 50, -1, 52,
       -1, 54, -1, 56, -1, 58, -1, 60, -1, 62, -1, 64, -1, 66, -1, 68, -1,
       70, -1, 72, -1, 74, -1])

### Упражнение 3

* Найдите в документации функцию, которая удаляет из одного массива элементы, которые есть в другом.  
* Приведите примеры использования.


Функцию удаления элементов, содержащихся в другом массиве выполняет numpy.delete(a, obj, axis=None)
Для определения наличия в первом массиве элементов другого массива можно использовать маску в виде второго массива.

### Упражнение 4

* Создайте случайную квадратную матрицу случайного размера от 10 до 100.
* Найдите максимум и сумму элементов.
* Поделите каждый элемент матрицы на максимум.

In [23]:
ex4_random_size = np.random.randint(10,100)
ex4_random_size

27

In [24]:
ex4_random_matrix = np.random.random((ex4_random_size, ex4_random_size))
ex4_random_matrix

array([[9.79320017e-01, 1.67406745e-01, 4.03619353e-01, 5.95155261e-01,
        8.10903300e-01, 5.77368757e-01, 1.28012452e-01, 7.05870888e-01,
        6.65279679e-01, 7.63388990e-01, 4.89337456e-01, 1.53278482e-02,
        6.91317582e-01, 8.59194748e-01, 2.60208575e-01, 9.84202602e-01,
        3.60665530e-01, 7.86473338e-01, 7.12595086e-01, 6.68155273e-03,
        3.51482817e-01, 1.30894802e-01, 6.95763987e-01, 1.47686272e-01,
        2.79322316e-01, 4.20831300e-01, 8.61610133e-01],
       [7.74310330e-01, 1.36779541e-01, 2.70932979e-01, 9.36660174e-01,
        9.24603837e-01, 1.73154169e-01, 9.80025278e-01, 7.43147537e-01,
        9.73807311e-01, 7.99942453e-01, 7.47324559e-01, 1.60264711e-01,
        8.16398198e-01, 1.93474415e-01, 5.73534837e-01, 5.56116227e-01,
        7.31809091e-01, 3.42654446e-01, 9.11174894e-01, 2.22637535e-01,
        4.81742807e-01, 3.26116455e-01, 1.26354200e-01, 1.51991782e-01,
        5.29934326e-01, 2.64157479e-01, 7.22481928e-01],
       [4.98634532e-01

In [48]:
ex4_random_matrix.max()

0.9999856936565784

In [49]:
ex4_random_matrix.sum()

3328.1424067607513

In [50]:
ex4_random_matrix/ex4_random_matrix.max()

array([[0.85747279, 0.72665867, 0.10275945, ..., 0.24821797, 0.20532455,
        0.63478229],
       [0.86521475, 0.13909395, 0.88947259, ..., 0.98031406, 0.43471478,
        0.21229145],
       [0.92026507, 0.81765848, 0.30253771, ..., 0.30647607, 0.1565218 ,
        0.4720576 ],
       ...,
       [0.83615532, 0.10924019, 0.4904885 , ..., 0.92652912, 0.04385088,
        0.06968908],
       [0.50521604, 0.65173373, 0.59193064, ..., 0.40204969, 0.54127396,
        0.2227834 ],
       [0.98312267, 0.15048116, 0.75263402, ..., 0.2754036 , 0.66716138,
        0.5846857 ]])

### Упражнение 5

* Напишите функцию, которая берет многомерный вещественный массив с пропущенными значениями (np.nan) и возвращает его копию с заполненными пропусками (заменять средним значением всех элементов массива). Если в массиве одни пропущенные значения, заполните их нулями.

In [51]:
def fillna_with_mean(numpy_array):
    if numpy_array.isnan().all():
        numpy_array = np.nan_to_num(numpy_array)
    else:
        mean = numpy_array.mean()
        numpy_array = np.nan_to_num(numpy_array, nan = mean)
    return numpy_array  

### Упражнение 6

* Напишите функцию которая нормализирует заданный numpy-массив, так чтобы его значения лежали в интервале $[0, 1]$.


In [ ]:
def normalize_numpy_array(numpy_array):
    return (numpy_array - numpy_array.min()) / numpy_array.ptp()

### Упражнение 7


* Запишите numpy-array в файл.
* Считайте numpy-array из файла.


In [ ]:
np.save('numpy_array', numpy_array)
np.load('numpy_array')

### Упражнение 8

* Создайте случайный массив
* Найдите в массиве элемент ближайший к данному


In [16]:
given_num = float(input('Задай число от одного до ста '))
ex8_array = np.random.rand(np.random.randint(5,500))
ex8_array = ex8_array * np.random.randint(1,1000)
copy_array = ex8_array/given_num
copy_array.sort()
elem_to_find = copy_array[copy_array >= 1][0]*given_num
elem_to_find

Задай число от одного до ста100


100.12718077841993

### Упражнение 9

* Напишите функцию, которая возвращает $n \times m$ целочисленную матрицу с заданным значением на границах матрицы и с заданным значением внутри

In [20]:
def matrix_n_m(n, m, value):
    matrix = np.full((n,m), value)
    return matrix

### Упражнение 10

* Напишите функцию, которая сортирует строки данной матрицы по значениям заданного столбца в порядке убывания.


In [75]:
def ex_10(matrix, col_number):
    return matrix[matrix[:,col_number].argsort()[::-1]]

In [73]:
array_to_test = np.random.rand(4,4)
array_to_test

array([[0.5666569 , 0.48857185, 0.36408236, 0.63412107],
       [0.40565248, 0.00684163, 0.32661431, 0.91772542],
       [0.6697652 , 0.67374471, 0.06842212, 0.18661195],
       [0.84252149, 0.40200196, 0.41484999, 0.72523236]])

In [76]:
print(ex_10(array_to_test, 3))

[[0.40565248 0.00684163 0.32661431 0.91772542]
 [0.84252149 0.40200196 0.41484999 0.72523236]
 [0.5666569  0.48857185 0.36408236 0.63412107]
 [0.6697652  0.67374471 0.06842212 0.18661195]]


### Упражнение 11

* Напишите функцию, которая создает двоичную матрицу шахматной доски с заданным shape

Пример 4x4:

```python,
[[0, 1, 0, 1],
 [1, 0, 1, 0],
 [0, 1, 0, 1],
 [1, 0, 1, 0]]```


In [120]:
def chess_square_matrix(n):
    chess_matrix = np.zeros((n,n), dtype = np.int64)
    rows,cols = np.where(chess_matrix == 0)
    for i in rows:
        for j in cols:
            if (i+j) % 2 == 0:
                chess_matrix[i,j] = 1
    return chess_matrix

In [121]:
chess_square_matrix(6)

array([[1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1]])

### Упражнение 12
* Напишите игру [life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life), используя массивы numpy.
* Визуализируйте интерактивно ее результаты.